File: TUnit.TestProject/ClassDataSourceDrivenTests.cs
using TUnit.Assertions;
using TUnit.Assertions.Extensions;
using TUnit.Core.Interfaces;
using TUnit.TestProject.Dummy;

namespace TUnit.TestProject;

public class ClassDataSourceDrivenTests
{
    [Test]
    [ClassDataSource<SomeAsyncDisposableClass>]
    public void DataSource_Class(SomeAsyncDisposableClass value)
    {
        // Dummy method
    }

    [Test]
    [ClassDataSource<SomeAsyncDisposableClass>]
    public void DataSource_Class_Generic(SomeAsyncDisposableClass value)
    {
        // Dummy method
    }

    [Test]
    [ClassDataSource<InitializableClass>]
    public async Task IsInitialized_With_1_ClassDataSource(InitializableClass class1)
    {
        await Assert.That(class1.IsInitialized).IsTrue();
    }

    [Test]
    [ClassDataSource<InitializableClass, InitializableClass>]
    public async Task IsInitialized_With_2_ClassDataSources(InitializableClass class1, InitializableClass class2)
    {
        await Assert.That(class1.IsInitialized).IsTrue();
        await Assert.That(class2.IsInitialized).IsTrue();
    }

    [Test]
    [ClassDataSource<InitializableClass, InitializableClass, InitializableClass>]
    public async Task IsInitialized_With_3_ClassDataSources(InitializableClass class1, InitializableClass class2, InitializableClass class3)
    {
        await Assert.That(class1.IsInitialized).IsTrue();
        await Assert.That(class2.IsInitialized).IsTrue();
        await Assert.That(class3.IsInitialized).IsTrue();
    }

    [Test]
    [ClassDataSource<InitializableClass, InitializableClass, InitializableClass, InitializableClass>]
    public async Task IsInitialized_With_4_ClassDataSources(InitializableClass class1, InitializableClass class2, InitializableClass class3, InitializableClass class4)
    {
        await Assert.That(class1.IsInitialized).IsTrue();
        await Assert.That(class2.IsInitialized).IsTrue();
        await Assert.That(class3.IsInitialized).IsTrue();
        await Assert.That(class4.IsInitialized).IsTrue();
    }

    [Test]
    [ClassDataSource<InitializableClass, InitializableClass, InitializableClass, InitializableClass, InitializableClass>]
    public async Task IsInitialized_With_5_ClassDataSources(InitializableClass class1, InitializableClass class2, InitializableClass class3, InitializableClass class4, InitializableClass class5)
    {
        await Assert.That(class1.IsInitialized).IsTrue();
        await Assert.That(class2.IsInitialized).IsTrue();
        await Assert.That(class3.IsInitialized).IsTrue();
        await Assert.That(class4.IsInitialized).IsTrue();
        await Assert.That(class5.IsInitialized).IsTrue();
    }
}

public class InitializableClass : IAsyncInitializer
{
    public Task InitializeAsync()
    {
        IsInitialized = true;
        return Task.CompletedTask;
    }

    public bool IsInitialized { get; private set; }
}

File: TUnit.TestProject.VB.NET/ClassDataSourceDrivenTests.vb
Imports System.Threading.Tasks
Imports TUnit.Assertions
Imports TUnit.Assertions.Extensions
Imports TUnit.Core
Imports TUnit.Core.Interfaces
Imports TUnit.TestProject.Library.Models


Public Class ClassDataSourceDrivenTests

    <Test>
    <ClassDataSource(GetType(SomeAsyncDisposableClass))>
    Public Sub DataSource_Class(value As SomeAsyncDisposableClass)
        ' Dummy method
    End Sub

    <Test>
    <ClassDataSource(GetType(SomeAsyncDisposableClass))>
    Public Sub DataSource_Class_Generic(value As SomeAsyncDisposableClass)
        ' Dummy method
    End Sub

    <Test>
    <ClassDataSource(GetType(InitializableClass))>
    Public Async Function IsInitialized_With_1_ClassDataSource(class1 As InitializableClass) As Task
        Await Assert.That(class1.IsInitialized).IsTrue()
    End Function

    <Test>
    <ClassDataSource(GetType(InitializableClass), GetType(InitializableClass))>
    Public Async Function IsInitialized_With_2_ClassDataSources(class1 As InitializableClass, class2 As InitializableClass) As Task
        Await Assert.That(class1.IsInitialized).IsTrue()
        Await Assert.That(class2.IsInitialized).IsTrue()
    End Function

    <Test>
    <ClassDataSource(GetType(InitializableClass), GetType(InitializableClass), GetType(InitializableClass))>
    Public Async Function IsInitialized_With_3_ClassDataSources(class1 As InitializableClass, class2 As InitializableClass, class3 As InitializableClass) As Task
        Await Assert.That(class1.IsInitialized).IsTrue()
        Await Assert.That(class2.IsInitialized).IsTrue()
        Await Assert.That(class3.IsInitialized).IsTrue()
    End Function

    <Test>
    <ClassDataSource(GetType(InitializableClass), GetType(InitializableClass), GetType(InitializableClass), GetType(InitializableClass))>
    Public Async Function IsInitialized_With_4_ClassDataSources(class1 As InitializableClass, class2 As InitializableClass, class3 As InitializableClass, class4 As InitializableClass) As Task
        Await Assert.That(class1.IsInitialized).IsTrue()
        Await Assert.That(class2.IsInitialized).IsTrue()
        Await Assert.That(class3.IsInitialized).IsTrue()
        Await Assert.That(class4.IsInitialized).IsTrue()
    End Function

    <Test>
    <ClassDataSource(GetType(InitializableClass), GetType(InitializableClass), GetType(InitializableClass), GetType(InitializableClass), GetType(InitializableClass))>
    Public Async Function IsInitialized_With_5_ClassDataSources(class1 As InitializableClass, class2 As InitializableClass, class3 As InitializableClass, class4 As InitializableClass, class5 As InitializableClass) As Task
        Await Assert.That(class1.IsInitialized).IsTrue()
        Await Assert.That(class2.IsInitialized).IsTrue()
        Await Assert.That(class3.IsInitialized).IsTrue()
        Await Assert.That(class4.IsInitialized).IsTrue()
        Await Assert.That(class5.IsInitialized).IsTrue()
    End Function

End Class

Public Class InitializableClass
    Implements IAsyncInitializer

    Public Property IsInitialized As Boolean

    Public Function InitializeAsync() As Task Implements IAsyncInitializer.InitializeAsync
        IsInitialized = True
        Return Task.CompletedTask
    End Function
End Class
